src/createMapper.ts

```diff
@@ -1,21 +1,43 @@
-import { Mapper, Profile } from '@/types'
+import { ProfileNotFoundError } from '@/profileNotFoundError'
+import { Mapper, Profile, ProfileKey } from '@/types'
 
 export function createMapper<T extends Profile>(profiles: T[]): Mapper<T> {
 
+  const mappers = buildProfilesMap(profiles)
+
+  function getMapper(sourceKey: string, destinationKey: string): Profile['map'] {
+    const key: ProfileKey = `${sourceKey}-${destinationKey}`
+    const mapper = mappers.get(key)
+
+    if (!mapper) {
+      throw new ProfileNotFoundError(sourceKey, destinationKey)
+    }
+
+    return mapper
+  }
+
   const mapper: Mapper<T> = {
     map: (sourceKey, source, destinationKey) => {
-      const profile = profiles.find(profile => profile.sourceKey === sourceKey && profile.destinationKey === destinationKey)
-
-      if (!profile) {
-        throw 'Mapping profile not found'
-      }
+      const mapper = getMapper(sourceKey, destinationKey)
 
-      return profile.map(source)
+      return mapper(source)
     },
     mapMany: (sourceKey, sourceArray, destinationKey) => {
-      return sourceArray.map(source => mapper.map(sourceKey, source, destinationKey))
+      const mapper = getMapper(sourceKey, destinationKey)
+
+      return sourceArray.map(source => mapper(source))
     },
   }
 
   return mapper
-}
\ No newline at end of file
+}
+
+function buildProfilesMap<T extends Profile>(profiles: T[]): Map<ProfileKey, Profile['map']> {
```

Comment:
should this be moved to a separate file?

Reply:
Maybe this should return `Map<ProfileKey, Profile>` and then in `createMapper` the utility for getting a mapper could be renamed to `getProfile` which reads better. `profile.map()` is nice as well.